### PR TITLE
Clarify backend and introduce "backend::none"

### LINKED ISSUE #564: Should we add "backend::none"?
I think we made a mistake in SYCL 2020 for `event::get_backend` in two cases:

* The default constructed `event`
* The `event` that is returned when you submit a host task

In both of these cases, there's no obvious backend.  As a result, there's no obvious backend interop that you can do with these events.  For example, what should `get_native` return for these events?

I think there's an implicit assumption that when `object.get_backend` returns `backend::opencl`, an application is allowed to use OpenCL backend interop APIs on that object.  Therefore, I think it was a mistake to return `backend::opencl` for events that are created in the two cases I list above.

I think the cleanest solution would be to introduce a new enumerator called `backend::none` for cases like this.  When an object reports `backend::none`, we would not support any backend interop for that object.

I also think `backend::none` would be useful for implementations that do not layer on any established "back end".  For example, you could imagine an implementation that implements SYCL directly on hardware, without going through any documented backend API.  Such an implementation could return `backend::none` for all calls to `get_backend`.

I think the specification should clearly specify what `backend` enumerator is returned, though.  Implementations should not have complete freedom to choose on an object-by-object basis.  Here are the rules that make sense to me:

* An implementation has complete freedom to decide what backend is returned from `platform::get_backend` for each `platform` object .  This can be a Khronos-defined enumerator like `backend::opencl`, a vendor-defined enumerator like `backend::ext_oneapi_level_zero`, or `backend::none`.  Each backend enumerator (except for `backend::none`) should have a matching backend interop specification.  Khronos will define this specification if the backend enumerator is defined by Khronos.  A vendor should define this specification if the backend enumerator is defined by the vendor.

* The SYCL specification should clearly document what is returned from all other `class::get_backend` member functions.  In most cases, we will say that it returns the same backend as the associated platform.  In the two "event" cases that I document above, we will say that `backend::none` is returned.

This seems like the right balance of implementation freedom and API consistency to me.

Since this affects behavior of SYCL 2020 APIs, I think we should consider this as a bug fix to the SYCL 2020 specification.

Comments?

### PR DESCRIPTION
**Pull request by gmlueck**
Clarify exactly what we mean by "backend".  I think it has always been our intention that the `backend` enumeration and the `get_backend` member function tell an application what sort of interoperation is supported by a SYCL object.  For example, if a SYCL object returns `backend::opencl` from its `get_backend` function, that SYCL object supports interoperation as defined by the OpenCL backend interoperation specification.

This commit clarifies the spec to say that explicitly.

It also became clear that we need a `backend::none` enumerator to indicate that a SYCL object does not support any backend interoperation. For example, this would be useful if a vendor implements SYCL directly on hardware, without using a documented low-level offload API.  Many of the changes in this commit merely remove sentences that imply that every SYCL object necessarily has some backend.

This commit also makes a minor change to the default `event` constructor.  Previously, we required the backend for such an event to be the same as the backend for the default device.  After some implementation experience, we decided that this is not practical. Requiring interoperation with a particular backend constrains the implementation too much.  It also seems arbitrary to require the default-constructed event to have a particular backend.  To address these concerns, this commit loosens the requirement, allowing an implementation to choose which backend (if any) the default-constructed event has.

Closes #564

### CODE REVIEW

File: adoc/chapters/architecture.adoc [hunk 1]
```diff
@@ -233,29 +219,37 @@ as SPIR-V, that will be finalized during execution or a final device ISA.
...
+Regardless of how the devices are exposed, they must implement the semantics of
+the SYCL APIs defined in this document.
+
+When a SYCL implementation layers on top of a heterogeneous API such as OpenCL,
+we refer to this as a <<backend>>.
+In this case, the implementation may choose to document how the backend APIs
+correlate to SYCL and provide interoperation APIs that allow a SYCL application
+to make calls to both the SYCL APIs and the backend APIs.
+However, applications that make use of this ability are no longer conformant to
+the core SYCL specification and may not be portable to other implementations of
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/architecture.adoc]: > Does the backend-api part of the Specification? do we define core SYCL specification somewhere?

Yes, "core SYCL specification" is defined in the glossary as:

> The text of the SYCL language specification (this document), excluding the text of any backend specifications and excluding the text for any extensions.

I updated the text here to use a link to the glossary term in 24292c4163d26559710eb2b162d568f435891d98

> I think we should be clearer that doing any call to the backend APIs is UB.

I don't think we should say this.  Instead, I think each backend interop specification should document what use of backend APIs produces defined behavior and what use produces UB.  I agree that the OpenCL backend interop specification could be improved in this regard, but fixing this is beyond the scope of this PR.

File: adoc/chapters/programming_interface.adoc [hunk 2]
```diff
@@ -15935,7 +15949,10 @@ The following member functions provide various queries for a <<kernel>>.
 backend get_backend() const noexcept;
 ----
 
-_Returns:_ The backend associated with this kernel.
+_Returns:_ The backend (if any) that underlies this object.
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc:15938]: I specified `backend::none` as "not a backend".  Here is the description from the section on the `backend` enumeration:

> The enumerator value `backend::none` is used to identify a SYCL object that does not correspond to any backend.

In addition, the section "The SYCL backend model" describes backends this way:

> A SYCL implementation may expose devices by layering on top of a heterogeneous API such as OpenCL or it may expose devices in some other way. [...] When a SYCL implementation layers on top of a heterogeneous API such as OpenCL, we refer to this as a SYCL backend.

Thus, when an implementation exposes devices without layering on top of a heterogeneous API, it is not using a backend.

File: adoc/chapters/architecture.adoc [hunk 3]
```diff
@@ -3,36 +3,25 @@
...
+CPU.
+
+Some terminology from SYCL derives from OpenCL.
+However, OpenCL is not a prerequisite for implementing SYCL.
+A SYCL implementation may layer on top of a heterogeneous programming API such
+as OpenCL, but this is not necessary.
+It is also possible to implement SYCL directly on hardware.
+
+A function object that can execute on a <<device>> is called a
+<<sycl-kernel-function>>.
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/architecture.adoc:24]: I felt like we needed to provide some example of a "heterogeneous programming API", otherwise no one would know what we mean by that term.  I used OpenCL as the example because that's what the previous text used.  I don't feel that strongly about also adding a reference to CUDA.  We can if we want, but I don't think it adds much extra clarity.  I do *not* think we should list every possible heterogeneous programming API here, though.

File: adoc/chapters/programming_interface.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (TApplencourt) [reply adoc/chapters/programming_interface.adoc:15938]: Would you happen to know the reason for such a choice for restricting backend to "heterogeneous API" ? 

 That a `TBB` or a `OpenMP` are not considered backends seems odd. It makes sense that they don't provide any interop, but I think the general usage of the word "backend" if more general.
 
 For example, in the SimSYCL paper, they use the term "AdaptiveCpp (OMP backend)" (https://dl.acm.org/doi/fullHtml/10.1145/3648115.3648136 table 2) who make total sense to me, but are not technically correct.
 
 In short, I don't understand the rationale behind this spec subtility. Every backend can decide if they provide or not some interopts for whatever SYCL function they want. 

PS: This also led to a question about an MPI Implementation. As far as I know, MPI is not a "heterogeneous API" (just an API :) ), but  I guess we still want to be able to give some interrupt handle.

File: adoc/chapters/architecture.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (TApplencourt) [reply adoc/chapters/architecture.adoc]: >I updated the text here to use a link to the glossary term in https://github.com/KhronosGroup/SYCL-Docs/commit/24292c4163d26559710eb2b162d568f435891d98

Oh Sorry I missed it when CTRL+F the spec! Thanks! 

>I don't think we should say this. Instead, I think each backend interop specification should document what use of backend APIs produces defined behavior and what use produces UB. I agree that the OpenCL backend interop specification could be improved in this regard, but fixing this is beyond the scope of this PR.

Make sense

File: adoc/chapters/architecture.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (TApplencourt) [reply adoc/chapters/architecture.adoc:24]: I agree; it's just we said explicitly that we are not bound to OpenCL and always use OpenCL as an example of a backend.  Just seem a little odd to me, but not important

File: adoc/chapters/programming_interface.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (illuhad) [reply adoc/chapters/programming_interface.adoc:15938]: > That a TBB or a OpenMP are not considered backends seems odd. It makes sense that they don't provide any interop, but I think the general usage of the word "backend" if more general.

Yes. I'm opposed to tying the notion of a "backend" to the ability of providing interop. Those two have been separate concepts so far in SYCL.

AdaptiveCpp treats OpenMP just like any other backend. I see no reason to change the backend model for "native CPU" approaches.

**Comment by illuhad** [comment]
>  I think it has always been our intention that the backend enumeration and the get_backend member function tell an application what sort of interoperation is supported by a SYCL object. For example, if a SYCL object returns backend::opencl from its get_backend function, that SYCL object supports interoperation as defined by the OpenCL backend interoperation specification.

I would be careful to tie the SYCL backend model too closely to the interop model. There may be other use cases for querying the backend, for example if you know that e.g. that a specific backend has certain performance characteristics.

I think am opposed to adding `backend::none`. It feels like it fundamentally changes the SYCL backend model, which is a core design aspect of SYCL, just to fix some edge cases of the interop model.

See discussion here for details: https://github.com/KhronosGroup/SYCL-Docs/issues/564

File: adoc/chapters/programming_interface.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc:15938]: The term "heterogeneous API" is used in the SYCL specification even before this PR.  As far as I know, there is no industry standard definition for this term.  I was assuming that OMP would also be considered a "heterogeneous API".

I'm not opposed to using a different term if someone has a suggestion.  Is there some better collective term for things like OpenCL, OpenMP, CUDA, etc.?  Maybe "offload API"?

[the remark this answers is not in the document]
Comment (TApplencourt) [reply adoc/chapters/programming_interface.adoc:15938]: I recommend disconnecting "heterogeneous API" from any "backend" consideration.   But maybe it is not the best place to discuss this. 

My main point is that we tied too much of the SYCL spec to the "heterogeneous API" of the backend.
For example:
>A function object that can execute on a [device](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#device) exposed by a [SYCL backend API](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#backend-api) is called a [SYCL kernel function](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sycl-kernel-function).

That means that for an OpenMP pure CPU or Intel TBB, a lambda used in Q.submit is not a kernel function. That also means that we can have "devices" that are not part of any "backend"?

OMP for offloading I I agree, is "heterogenous"; for CPU only OpenMP I don't see why it's "heterogeneous".  Same for MPI. Or for a `pthread` CPU implementation of SYCL. 

I think all this "SYCL is build on top of a backend API" is a relic of OpenCL; we should clarify that it's just implementation details. SYCL provides a "heterogenous API" but doesn't need to one be implemented. The spec should be clear about it.

And I think the easiest way is to define loosely what is a backend; it's whatever the implementation wants it to be, and not tie it to any concept of "heterogenous API"

But yeah, maybe a far more broad conversation; sorry for the side-discussion.

**Comment by tomdeakin** [comment]
Next steps:
1. PR for "Default constructed event should be associated with an implementation defined backend."
2. More discussion on backend and interop

**Comment by TApplencourt** [comment]
To clarify my point (sorry if I was unclear earlier), we have two distinct types of "interop errors."  

- One is backend-specific. The backend implementer can decide what they implement and what to not implement (but still provide the API). For example, `get device interopt`  for  CPU only or whatever.
This is a local decision made by the backend implementer. And should be part of 

- The second is a "cross-backend" error.  For example, a default constructed event is not associated with any backend. The question is not about the "interopt" but about which backend this default event is bound to.  This cannot be solved by any backend-spec, it need to be specified inside the core-sycl-spec (and as we reach said it's UB).


One is a "I Don't Know with Backend It's sorry, general SYCL problem" and the other is a "Not Implement Error for this backend. Contact your vendor.".  
Using the return value of "backend" for those 2 errors, seem wrong to me. They are solved in the same part of the spec.  One is a general SYCL UB problem, the other an "implementation" detail.

For me `backend::none` should be reserved for the "cross backend problem". The other non-implemented / non valid usage of interopt API should `throw`.  If we want to avoid the `throw,`, we should add a new API call, something like `is_get_native_avalaible` or whatever  that people can use to check if they can use `get_native` or not.
 

Hope this clarify,

**Comment by gmlueck** [comment]
> PR for "Default constructed event should be associated with an implementation defined backend."

I created a separate PR with just this change #582

File: adoc/chapters/architecture.adoc [hunk 4]
```diff
@@ -306,22 +299,21 @@ across <<backend, SYCL backends>>.
...
+The SYCL programming model exposes a number of <<platform,platforms>>, each of
+them either empty or exposing a number of <<device,devices>>.
+Each platform is bound to no more than one <<backend>>.
+If the platform is bound to a backend, the devices associated with that platform
+are associated with that backend.
 
-Although the APIs in the SYCL generic programming model are defined according to
-this specification and their version is indicated by the macro
+Although most APIs in the SYCL programming model are defined according to this
+specification and their version is indicated by the macro
```
Comment (keryell) [inline adoc/chapters/architecture.adoc:316]: Even if it was in the previous version, `and their version` looks weird to me.
Perhaps we can clarify the meaning?

**Review by keryell — APPROVED** [review]
That sounds good.

File: adoc/chapters/architecture.adoc (hunk 4 above)
Comment (gmlueck) [reply adoc/chapters/architecture.adoc:316]: If we want to clean this up, I would propose deleting this entire paragraph.  I don't think it adds any clarity.  I don't think the core spec needs to say anything about the versioning of the backend interoperation APIs.  This can be left to each backend interoperation specification.
